File: java_solutions/1376.time-needed-to-inform-all-employees.java
/*
 * @lc app=leetcode id=1376 lang=java
 *
 * [1376] Time Needed to Inform All Employees
 */

class Solution {
    public int numOfMinutes(int n, int headID, int[] manager, int[] informTime) {
        return java.util.Optional.of(new java.util.List[n])
            .filter(adj -> java.util.stream.IntStream.range(0, n).allMatch(i -> (adj[i] = new java.util.ArrayList<Integer>()) != null))
            .filter(adj -> java.util.stream.IntStream.range(0, n).allMatch(i -> manager[i] == -1 || adj[manager[i]].add(i)))
            .map(adj -> new Object[]{adj, new java.util.function.IntUnaryOperator[1]})
            .map(s -> (((java.util.function.IntUnaryOperator[]) s[1])[0] = id -> informTime[id] + (((java.util.List<Integer>[]) s[0])[id].isEmpty() ? 0 : ((java.util.List<Integer>[]) s[0])[id].stream().mapToInt(c -> ((java.util.function.IntUnaryOperator[]) s[1])[0].applyAsInt(c)).max().getAsInt())).applyAsInt(headID))
            .orElse(0);
    }
}

File: kotlin_solutions_1/1376.time-needed-to-inform-all-employees.kt
/*
 * @lc app=leetcode id=1376 lang=kotlin
 *
 * [1376] Time Needed to Inform All Employees
 */

class Solution { fun numOfMinutes(n: Int, headID: Int, manager: IntArray, informTime: IntArray): Int = Array(n) { mutableListOf<Int>() }.apply { manager.indices.forEach { i -> if (manager[i] != -1) this[manager[i]].add(i) } }.let { adj -> DeepRecursiveFunction<Int, Int> { u -> informTime[u] + (adj[u].maxOfOrNull { callRecursive(it) } ?: 0) }.invoke(headID) } }